bootstrap/kubeadm/controllers/kubeadmconfig_controller.go

```diff
@@ -524,13 +519,9 @@ func (r *KubeadmConfigReconciler) joinControlplane(ctx context.Context, scope *S
 	}
 	conditions.MarkTrue(scope.Config, bootstrapv1.CertificatesAvailableCondition)
 
-	// ensure that joinConfiguration.Discovery is properly set for joining node on the current cluster
-	if err := r.reconcileDiscovery(ctx, scope.Cluster, scope.Config, certificates); err != nil {
-		if requeueErr, ok := errors.Cause(err).(capierrors.HasRequeueAfterError); ok {
-			scope.Info(err.Error())
-			return ctrl.Result{RequeueAfter: requeueErr.GetRequeueAfter()}, nil
-		}
-		return ctrl.Result{}, err
+	// Ensure that joinConfiguration.Discovery is properly set for joining node on the current cluster.
+	if res, err := r.reconcileDiscovery(ctx, scope.Cluster, scope.Config, certificates); err != nil || !res.IsZero() {
+		return res, err
 	}
```

Comment:
This feels a bit odd given the convention of not relying on other values when an error occurs, might be better to be explicit here:
```suggestion
	res, err := r.reconcileDiscovery(ctx, scope.Cluster, scope.Config, certificates)
	if err != nil  {
		return ctrl.Result{}, err
	}
	if !res.IsZero() {
		return res, nil
	}
```